### DISCUSSION TITLE
Future Changes & Features

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/aurelia-slickgrid#490] [Ideas]
Now that Aurelia-Slickgrid `3.x` is released and out of the way, I can list here the potential focus of upcoming changes/features.

_Some of them might be considered as ideas only, so it doesn't mean that they will be all implemented_
_Also note that many of these ideas will be implemented in Slickgrid-Universal and will bubble up to Aurelia-Slickgrid_

- [x] 1. Add "Clone Item/Row" feature to the Composite Editor (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/236))
- [x] 2. Replace MomentJS to [date-fns](https://github.com/date-fns/date-fns) which seem to be the best alternative to Moment, it supports TypeScript, is Tree Shakable and is much smaller in size. (some simple migration [guide 1](https://gist.github.com/nickscript0/adbbfe697c824b70a0399d57e51ad043), [guide 2](https://jmduke.com/2017/10/07/migrating-from-moment-to-date-fns), [guide 3](https://dev.to/mullinsmikey/farewell-moment-js-bd3))
   - or go with [Day.js](https://day.js.org/en/) since it's very close to MomentJS API (missing the Zulu "Z" in strict mode is the biggest blocking issue)
   - none of the libs I tried were working out of the box and I think I will end up waiting for [Temporal](https://tc39.es/proposal-temporal/docs/index.html) to be out (currently at Stage 3)
   - finally migrating to Tempo and getting rid of Moment (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/1507))
- [x] 3. Dark Styled Themes 
   - mostly just an idea that I always had in mind, it would be nice to have help on this subject
- [x] 4. Migrate Slickgrid-Universal from ~Lerna to NPM Workspaces~
  - NPM Workspaces aren't supporting script in parallel & in topological order, so I ended up forking Lerna and created [Lerna-Lite](https://github.com/ghiscoding/lerna-lite) which I now use
- [x] 5. Migrate Slickgrid-Universal/Aurelia-Slickgrid from CircleCI to GitHub Actions
   - [x] add an Action to auto-publish GitHub Demo page after each build
   - [x] apply for Cypress OSS plan 
- [x] 6. Upgrade demo to WebPack 5 (requires Aurelia WebPack 5 [issue](https://github.com/aurelia/aurelia/issues/982) to be resolved)
   - [x] Aurelia-Slickgrid to WebPack 5
   - [x] Slickgrid-Universal to WebPack 5 (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/225))
- [x] 7. Upgrade internals and demos to Aurelia 2.x (whenever that gets closer to a release) (ref PR #1038)
- [x] 8. Should we support RxJS?  (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/280))
    - Angular-Slickgrid is built with RxJS, because that is what they use, so I wonder if it make sense to add it into Slickgrid-Universal?
    - Perhaps all I need is an abstract class/methods, which would return false internally, that would be implemented in the targeted framework like Angular-Slickgrid
 - [x] 9. (WON'T DO - Leave as is) Use [TypeDI](https://docs.typestack.community/typedi/v/develop/01-getting-started) for agnostic DI to replace manual instantiation within the lib. 
    - Not sure of a real benefit in doing that though, it might increase the build size, so I might never implement it. Just adding it here as a reminder.
 - [x] 10. Flatpickr Locale seems to all be loaded in the final bundle, need to investigate and preferably find better alternative to loading different locale (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/227))
 - [x] 11. Support Bootstrap 5 (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/226))
    - [x] add new BS5 demo as well
    - [x] post BS5 official release, let's remove the BS3 demo, make BS4 the default demo and BS5 as the aurelia-slickgrid-demos default
 - [ ] 12. PDF Export, see UI-Grid [data export](http://ui-grid.info/docs/#!/tutorial/Tutorial:%20206%20Exporting%20Data) and in [code](https://github.com/angular-ui/ui-grid/blob/master/packages/exporter/src/js/exporter.js#L135)
 - [x] 13. add SlickGrid Aggregators to Tree Data (done in Slickgrid-Universal [PR 1074](https://github.com/ghiscoding/slickgrid-universal/pull/1074))
 - [x] 14. change TSC builds from ES5 to ES2018 (ref PR #566)
 - [x] 15. auto-resize columns by their cell content (ref PR #560)
- [x] 16. move all SlickGrid controls/plugins directly into Slickgrid-Universal with proper unit tests (ref [PR](https://github.com/ghiscoding/slickgrid-universal/pull/555))
- [x] 17. remove jQueryUI from SlickGrid core dependencies (ref PR #865)
- [x] 18. remove jQuery from SlickGrid core dependencies

Again this is more of a wish list than anything else but I'm open to suggestions... and totally open for contributions 👷 
Cheers 🍺 ☕ 

The most likely candidate to soon be looked into would be ~number 2~ and number 5

This library was created over 3 years ago... time flies 🚀

**Comment by zewa666** [comment]
Date-fns is a good choice, had no probs with it either.

3: always nice to have ;) 
5: good idea. Enjoying GitHub Actions much more than CircleCI

7: soon ;)

8: ng needs it anyways for every other framework its just an additional deps burden. So I wouldnt do that unless absolutely needed.

9: feels heavy for this libs use case

**Reply by ghiscoding** [reply]
If you have a good GitHub Actions setup to build TSC, run Jest and Cypress, I'd be happy to receive a PR or a reference link. Because, the only time I've tried it so far was to add it into [6pac/SlickGrid](https://github.com/6pac/SlickGrid) repo to run Cypress and that is all the experience I have so far.

**Reply by zewa666** [reply]
Its close after midnight for you so merry XMas ;)

Correct me if I'm wrong but the CI run would essentially just call a series of npm scripts right? Havent yet looked into your circle conf. Do you have any other special requirements like handling the artifacts?

**Reply by ghiscoding** [reply]
lol yes it's just past midnight but now it's the 24, we are behind not forward, so not yet the 25th ;) 

and yes that would probably be a bunch of npm scripts which you can see in the CircleCI [config](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/.circleci/config.yml), however I would try to use Actions when available, for example with Cypress I found this one `cypress-io/github-action@v2` but that is actually the part that confuses me, I'm not sure when to use the configured actions or just run just a bunch of scripts, I think with Cypress it make sense because it has a bunch of setup but I don't know about others (TSC, Jest) I guess npm scripts would be just fine for them?! I never look, neither care about previous artifact, so we can skip that part.

All I care is.. build TSC, run Jest tests+upload to codecov, run WebPack Dev & run Cypress tests... with CircleCI, I was doing all of that in 1 pass and I didn't need any artifacs.

**Reply by zewa666** [reply]
So the idea is that those custom workflows/actions often provide convenience features. In the case of Cypress it has stuff like artifact handling, running a Dev Server in the background and others. For TSC and Jest though there wouldnt be too much of a benefit so simple npm scripts are fine. I guess you do codecov with npm script hooks so that is also cared for.

**Reply by zewa666** [reply]
Oh and Christmas Eve is celebrated on the evening of 24th here in Austria, so people kinda start the celebration already today. To make things more complicated lots of good friends being orthodox christians celebrate it on 7th of Jan. I guess there were some competing forks back then and nobody cared to do a forward merge 🤣

**Reply by ghiscoding** [reply]
in SlickGrid, I had Cypress working with the simple http-server package, in my case I just need to replace that with WebPack Dev Server (via npm scripts maybe). Here's the config I created on SlickGrid, 
```yml
- name: Run Cypress E2E tests
        uses: cypress-io/github-action@v2
        with:
          start: npm run serve
          wait-on: 'http://localhost:8080'
          browser: chrome
```

**Reply by zewa666** [reply]
Yep it should be as simple as that. I saw though in circle you're copying things to a different Reports folder, but guess that was about codecov

**Reply by ghiscoding** [reply]
lol right, it's been a long time since I've celebrated on Christmas eve but yeah I know the customs... it's also a good time to give Santa a cup of milk with a cookie hehe... bedtime for me now, ttyl, enjoy Xmas

**Comment by ghiscoding** [comment]
@zewa666 
I got a question for you since you also use multiple translations for some projects. 
So I found out, number 10 in the list, that the entire set of Flatpickr (date picker) Locales (60+) were being bundled up because I was dynamically loading the locale via `require('flatpickr/dist/l10n/' + language + '.js').default;` but that is not tree shakable so everything gets bundled up.

So the best alternative I could come up with was to ask the user to import whichever locale he wants to use, for example if I want to use French then I would load `import 'flatpickr/dist/l10n/fr';` and in case the user forgot to do it, then I do this console warning
```ts
console.warn(`[Slickgrid-Universal] Flatpickr missing locale imports, will revert to English as the default locale.
          See Flatpickr Localization for more info, for example if we want to use French, then we can import it with:  import 'flatpickr/dist/l10n/fr';`);
```

As a user, would you be ok with that? 
It's not really a breaking change, but it's an extra step. 
I just couldn't find a better alternative and I really want to get rid of these 60+ locales (month & days, short/long texts) from my bundle.

More info on the [Flatpickr - Localization](https://flatpickr.js.org/localization/)

EDIT

I think that is the best option and so I made the [PR](https://github.com/ghiscoding/slickgrid-universal/pull/227) for the change

EDIT 2

Loading it in the `app.html` file seems to work, is that the best way? See this fix [commit](https://github.com/ghiscoding/aurelia-slickgrid-demos/commit/42fb7c7db8a9428eacbf5b9231e6797ec320312e) 
```html
<require from="bootstrap/css/bootstrap.css"></require>
<require from="https://npmcdn.com/flatpickr/dist/l10n/fr.js"></require>
```
